### DISCUSSION TITLE
Getting a discrepancy when using to_undirected()

### DISCUSSION THREAD

**Discussion by scigeek72** [pyg-team/pytorch_geometric#5496] [General]
Hello,

I am building a graph data from a .csv file. Once the data is built (it is directed at this point), I am using `to_undirected` to convert it to an `undirected` graph. Below is what I am getting depending on whether I am using `num_nodes` parameter in `to_undirected`.

```
c = to_undirected(edge_index=ei_tr, edge_attr=ea_tr)

c[0].shape
Out[224]: torch.Size([2, 33282])

is_undirected(c[0])
Out[225]: True
```
now with `num_nodes`

```
n_nodes_tr = 3786
d = to_undirected(edge_index=ei_tr, edge_attr=ea_tr, num_nodes= n_nodes_tr)

d[0].shape
Out[227]: torch.Size([2, 33258])

is_undirected(d[0])
Out[228]: False


```
Note that the number of directed edges is 16641. So, without `num_nodes`, the code seems to produce the correct shape.

When I am using `num_nodes`, the shape of the `edge_index` changes (my guess is that this is causing `is_undirected()` to be `False` as for some edges, the inverse is not getting included in the output of the function). 

I looked at the source code for `to_undirected()` ([here](https://github.com/pyg-team/pytorch_geometric/blob/25459b4812db7a8e2abe30191f91017771d60640/torch_geometric/utils/undirected.py#L67)) but couldn't resolve this behavior.  

Can `num_nodes` change the behavior of `to_undirected()`? 

Your help is greatly appreciated. Thanks.

**Comment by EdisonLeeeee** [comment]
Hi @scigeek72 
In your final example, I guess `num_nodes` is smaller than the actual number of nodes in your graph, i.e., `num_nodes < ei_tr.max() + 1`, which results in unexpected behaviors.

**Reply by scigeek72** [reply]
Thanks for your reply @EdisonLeeeee, but that bothers me a bit (although I may be wrong). `num_nodes == ei_tr.max()+1` may not always be true, for example, if the graph has isolated nodes. We may not have that information prior to building the dataset. Certainly, in my case, I need to first build the graph and `pytorch_geometric` data before I know that there are `isolated` nodes. 

Also, you are assuming that nodes are renamed between `0` and `range(len(number of nodes))`. But is it a fair assumption? 

Again, my view of the situation is through my own problem. May be in the larger context, your assumptions are correct. 

Thanks for taking the time to respond.

**Reply by EdisonLeeeee** [reply]
> num_nodes == ei_tr.max()+1 may not always be true, for example, if the graph has isolated nodes. 

I agree with that there are cases with isolated nodes would result in unexpected behaviors. In that cases, PyG assumes that users must be responsible for their inputs and they should make sure `num_nodes > max_node_id`. One can use `torch_geometric.utils.remove_isolated_nodes` to remove isolated nodes and reorder them.
```python
edge_index = ... # build graph
edge_index = remove_isolated_nodes(edge_index)[0]
edge_index = to_undirected(edge_index)[0]
```
> Also, you are assuming that nodes are renamed between 0 and range(len(number of nodes)). But is it a fair assumption?

In fact, nodes should be renamed between `0` and `range(max_node_id)`, so PyG uses `num_nodes == ei_tr.max()+1` to figure out the number of nodes by default. Not ideally, but that would make things simple in most standard cases.
